pkg/pmem-csi-operator/controller/deployment/controller_driver.go

```diff
@@ -198,23 +219,152 @@ func (d *PmemCSIDriver) reconcileDeploymentChanges(r *ReconcileDeployment, chang
 		}
 	}
 
+	obsoleteList, err := d.getObsoleteObjects(r.operatorVersion)
+	if err != nil {
+		klog.Warningf("Failed to get obsolete objects for deployment %q: %v", d.Name, err)
+	}
+	for _, o := range obsoleteList {
+		if err := r.Delete(o); err != nil && !errors.IsNotFound(err) {
+			klog.Warningf("Failed to delete obsolete object %T for deployment %q: %v", o, d.GetName(), err)
+		}
+	}
+
+	if err := d.tagOwnedObjects(r.operatorVersion); err != nil {
+		klog.Warningf("Internal error while updating the annotation %s: %v", d.GetName(), err)
+	}
+
 	return false, nil
 }
 
-func (d *PmemCSIDriver) deployObjects(r *ReconcileDeployment) error {
-	objects, err := d.getDeploymentObjects()
-	if err != nil {
-		return err
+// tagOwnedObjects updates the deployment's annotation with the information(object name:GVK)
+// of the objects it owned as a json string.
+func (d *PmemCSIDriver) tagOwnedObjects(operatorVersion string) error {
+	nameGVKs := []string{}
+
+	// Nothing to update if this deployment is created by the same operator version
+	if d.ObjectMeta.Annotations[versionAnnotation] == operatorVersion {
+		return nil
 	}
-	for _, obj := range objects {
-		if err := r.Create(obj); err != nil {
+
+	for _, obj := range d.getDeploymentOwnedObjects() {
+		metaObj, err := meta.Accessor(obj)
+		if err != nil {
 			return err
 		}
+		gvk := obj.GetObjectKind().GroupVersionKind().String()
+		nameGVKs = append(nameGVKs, metaObj.GetName()+":"+gvk)
 	}
+
+	jsonString, err := json.Marshal(nameGVKs)
+	if err != nil {
+		return err
+	}
+
+	d.ObjectMeta.Annotations[versionAnnotation] = operatorVersion
+	d.ObjectMeta.Annotations[ownedObjectsAnnotation] = string(jsonString)
+	d.needsUpdate = true
+
 	return nil
 }
 
-// getDeploymentObjects returns all objects that are part of a driver deployment.
+// retrieveTaggedObjects retrives the objects tagged for this deployment by the
```

Comment:
"retrieveTaggedObjects retrives the objects tagged" just repeats the same information. When describing a function, at least try to paraphrase or (better) add additional information. In this case, when reading just this function name and comment, it is not clear what "tagged" means.